src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,304 @@
+        debug!("visit_foldable: value={:?}", value);
+        // Exit early if this type has no type or const parameters.
+        if !value.has_param_types_or_consts() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any generic parameters.
+        value.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let is_closure_or_generator =
+            self.tcx.is_closure(self.def_id) || self.tcx.type_of(self.def_id).is_generator();
+        if is_closure_or_generator && local == Local::from_usize(1) {
+            // Skip visiting the closure/generator that is currently being processed. This only
+            // happens because the first argument to the closure is a reference to itself and that
+            // will call `visit_substs`, resulting in each generic parameter captured being
+            // considered used by default.
+            debug!("visit_local_decl: skipping closure substs");
+            return;
+        }
+
+        self.super_local_decl(local, local_decl);
+    }
+
+    fn visit_const(&mut self, c: &&'tcx Const<'tcx>, _: Location) {
+        self.visit_foldable(*c);
+    }
+
+    fn visit_ty(&mut self, ty: Ty<'tcx>, _: TyContext) {
+        self.visit_foldable(ty);
+    }
+}
+
+impl<'a, 'tcx> TypeVisitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_const(&mut self, c: &'tcx Const<'tcx>) -> bool {
+        debug!("visit_const: c={:?}", c);
+        match c.val {
+            ty::ConstKind::Param(param) => {
+                debug!("visit_const: param={:?}", param);
+                set_bit(self.used_parameters, param.index);
+                false
+            }
+            _ => c.super_visit_with(self),
+        }
+    }
+
+    fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+        debug!("visit_ty: ty={:?}", ty);
+        match ty.kind {
+            ty::Closure(def_id, _) | ty::Generator(def_id, ..) => {
+                debug!("visit_ty: def_id={:?}", def_id);
+                // Avoid cycle errors with generators.
+                if def_id == self.def_id {
+                    return false;
+                }
+
+                // Consider any generic parameters used by any closures/generators as used in the
+                // parent.
+                let results = self.tcx.unused_generic_params(def_id);
+                debug!(
+                    "visit_ty: used_parameters={:064b} results={:064b}",
+                    self.used_parameters, results
+                );
+                // Can just `|` the results of running the analysis on the closure/generator, the
+                // synthetic parameters will be marked as used and be the rightmost bits, so
+                // won't make a different.
+                *self.used_parameters |= !results;
```

Comment:
This only makes sense when the closure/generator type shows up in the parent, right? I think a "more correct" implementation would do this:

```rust
let unused = self.tcx.unused_generic_params(def_id);
for (i, arg) in substs.iter().enumerate() {
    let is_unused = i < 64 && (unused & (1 << i)) != 0;
    if !is_unused {
        arg.visit_with(self);
    }
}
```

This is effectively "masking" the closure/generator substs so that only those that are used get visited.

It will be slightly slower for the typical case of e.g. the parent function of a closure, but it's also correct even if e.g. an `-> impl Fn()` got revealed, or if MIR inlining ran (did you ever test with MIR inlining turned on?)

Reply:
Fixed!